src/main/java/com/solbeg/Main.java

```diff
@@ -45,30 +50,36 @@ public List<User> findUsersByBirthdayMonth(Month birthdayMonth) {
      * @return Map<String, List < User>> where key is email domain and value is a list of users
      */
     public Map<String, List<User>> groupUsersByEmailDomain() {
-        throw new ExerciseNotCompletedException();
+        return users.stream()
+                .collect(Collectors.groupingBy(user -> user.getEmail().split("@")[1]));
     }
 
 
     /**
      * @return total balance of all users
      */
     public BigDecimal calculateTotalBalance() {
-        throw new ExerciseNotCompletedException();
+        return users.stream()
+                .map(User::getBalance)
+                .reduce(BigDecimal.ZERO, BigDecimal::add);
     }
 
     /**
      * @return list of users sorted by first and last names
      */
     public List<User> sortByFirstAndLastNames() {
-        throw new ExerciseNotCompletedException();
+        return users.stream()
+                .sorted(Comparator.comparing(User::getFirstName).thenComparing(User::getLastName))
+                .toList();
     }
 
     /**
      * @param emailDomain - for example "gmail.com"
      * @return true if there is a user that has an email with provided domain
      */
     public boolean containsUserWithEmailDomain(String emailDomain) {
-        throw new ExerciseNotCompletedException();
+        return users.stream()
+                .anyMatch(user -> user.getEmail().split("@")[1].equals(emailDomain));
```

Comment:
first add filter for email check
then reverse condition
emailDomain.equals(user.getEmail().split("@")[1])